osquery/events/windows/ntfs_event_publisher.cpp

```diff
@@ -0,0 +1,593 @@
+  if (!status.ok()) {
+    return status;
+  }
+
+  // See if we have been requested to solve the root folder
+  if (ref == volume_data.root_ref) {
+    path.push_back(drive_letter);
+    path.append(":\\");
+
+    return Status(0);
+  }
+
+  // Convert the reference number to the native Windows structure
+  FILE_ID_DESCRIPTOR native_file_id = {};
+  GetNativeFileIdFromUSNReference(native_file_id, ref);
+
+  // Attempt to open the file or folder
+  auto handle =
+      OpenFileById(volume_data.volume_handle,
+                   &native_file_id,
+                   0,
+                   FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+                   nullptr,
+                   FILE_FLAG_BACKUP_SEMANTICS);
+  if (handle == INVALID_HANDLE_VALUE) {
+    std::stringstream message;
+    message << "Failed to open the file in volume " << drive_letter
+            << ":\\. Error: ";
+
+    std::string description;
+    if (!getWindowsErrorDescription(description, ::GetLastError())) {
+      description = "Unknown error";
+    }
+
+    message << description;
+    return Status(1, message.str());
+  }
+
+  auto required_bytes = static_cast<size_t>(::GetFinalPathNameByHandle(
+      handle, nullptr, 0, FILE_NAME_NORMALIZED | VOLUME_NAME_DOS));
+
+  if (required_bytes == 0U) {
+    auto error_code = ::GetLastError();
+    ::CloseHandle(handle);
+
+    std::stringstream message;
+    message << "Failed to determine the path size for the file in volume "
+            << drive_letter << ":\\. Error: ";
+
+    std::string description;
+    if (!getWindowsErrorDescription(description, error_code)) {
+      description = "Unknown error";
+    }
+
+    message << description;
+    return Status(1, message.str());
+  }
+
+  // We are going to add an additional byte, as we may or may not have the null
+  // terminator already included depending on the operating system version
+  std::string buffer;
+  required_bytes += 1U;
+
+  buffer.resize(required_bytes);
+  if (buffer.size() != required_bytes) {
+    throw std::bad_alloc();
+  }
+
+  auto bytes_returned = static_cast<size_t>(
+      ::GetFinalPathNameByHandle(handle,
+                                 &buffer[0],
+                                 static_cast<DWORD>(buffer.size()),
+                                 FILE_NAME_NORMALIZED | VOLUME_NAME_DOS));
+
+  auto error_code = ::GetLastError();
+  ::CloseHandle(handle);
+
+  if (bytes_returned == 0U || bytes_returned >= buffer.size()) {
+    std::stringstream message;
+    message << "Failed to acquire the path for the file in volume "
+            << drive_letter << ":\\. Error: ";
+
+    std::string description;
+    if (!getWindowsErrorDescription(description, error_code)) {
+      description = "Unknown error";
+    }
+
+    message << description;
+    return Status(1, message.str());
+  }
+
+  // Paths follow this form: \\?\C:\\path\\to\\folder; skip the prefix
+  path = buffer.c_str() + 4;

```

Comment:
I don't *think* so -- the [`GetFinalPathNameByHandle`](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfinalpathnamebyhandlea) docs indicate that the returned path is always `\\?\`-prefixed, except for in the error case, where the returned length will either be `0` or greater than the supplied buffer size (both of which we handle above).

So, assuming that we pass the error checks, `buffer.size()` should always be `> 4` (and not just `>=`, since an empty path would be an invalid result).